src/CachedKeyDecoder.ts

```diff
@@ -0,0 +1,71 @@
+import { utf8DecodeJs } from "./utils/utf8";
+
+interface KeyCacheRecord {
+  readonly bytes: Uint8Array;
+  readonly value: string;
+}
+
+const DEFAULT_MAX_KEY_LENGTH = 16;
+const DEFAULT_MAX_LENGTH_PER_KEY = 32;
+
+export class CachedKeyDecoder {
+  private readonly caches: Array<Array<KeyCacheRecord>>;
+
+  constructor(
+    private readonly maxKeyLength = DEFAULT_MAX_KEY_LENGTH,
+    private readonly maxLengthPerKey = DEFAULT_MAX_LENGTH_PER_KEY,
+  ) {
+    // avoid `new Array(N)` to create a non-sparse array for performance.
+    this.caches = [];
+    for (let i = 0; i < this.maxKeyLength; i++) {
+      this.caches.push([]);
+    }
+  }
+
+  public canBeCached(byteLength: number) {
+    return byteLength > 0 && byteLength <= this.maxKeyLength;
+  }
+
+  private get(bytes: Uint8Array, inputOffset: number, byteLength: number): string | null {
+    const records = this.caches[byteLength - 1];
+    const recordsLength = records.length;
+
+    FIND_CHUNK: for (let i = 0; i < recordsLength; i++) {
+      const record = records[i];
+
+      for (let j = 0; j < byteLength; j++) {
+        if (record.bytes[j] !== bytes[inputOffset + j]) {
+          continue FIND_CHUNK;
+        }
+      }
+      return record.value;
+    }
+    return null;
+  }
+
+  private store(bytes: Uint8Array, value: string) {
+    const records = this.caches[bytes.length - 1];
+    const record: KeyCacheRecord = { bytes, value };
+
+    if (records.length >= this.maxLengthPerKey) {
+      // `records` are full!
+      // Set `record` to a randomized position.
+      records[(Math.random() * records.length) | 0] = record;
```

Comment:
I have replaced the original hit-count based sorting algorithm by random-picking algorithm to reduce overhead in `#get()`.